### PR TITLE
 Overhaul upload requests, Blob and DataPart

### LINKED ISSUE #344: How to upload customed Blobs (like DataPart) in multipart way?
I have a Blob list to upload with different names and filenames like

`listOf(blob("a","a1.jpg",size,stream),blob("a","a2.jpg",size,stream),
blob("b","b1.jpg",size,stream),blob("b",b2.jpg",size,stream))`

are there any way to do it with fuel like DataParts?

`DataPart(Blob(),"name", "image/jpeg")`

 or is this right for this situation

```kotlin
fun compareApi(faceImgData: FaceImgData,callback: (Result<CompareCBData, FuelError>)->Unit){
        compare_suffix.httpUpload(Method.POST).blob { request, url ->
            faceImgData.fullImg
        }.name {
            "fullImg"
        }.blobs { request, url ->
            faceImgData.faces
        }.name {
            "faces"
        }.responseObject(CompareCBData.Deserializer()) { request, response, result ->
            callback(result)
        }
    }
```

### LINKED ISSUE #390: Upload list of files as single element in multipart body
See the issue https://github.com/kittinunf/Fuel/issues/169, which seams try to get the same, but was closed before somebody give the solution.

I tried solution in that issue, but the resulting multipart body still contains separate part for each file. The name of the parts are "files1", "files2", etc.
What I need (and seams the originator of that topic) the multipart with single part named "files" and containing the list of files.

How can we do THAT?

### LINKED ISSUE #399: multipart/form-data boundary isn't set correctly
```
val result = FuelManager()
  .request(Method.PUT, "/", listOf("lat" to "3"))
  .apply { type = Request.Type.UPLOAD }
  .dataParts { request, url ->
    listOf(DataPart(File("some.jpg")))
  }.awaitStringResult()
```
This made some trouble with ktor server.
```
java.io.IOException: Failed to parse multipart: Content-Type should be multipart/* but it is application/x-www-form-urlencoded
```
[`Request.dataParts`](https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Request.kt#L182) doesn't seem set boundary header, and [`UploadTaskRequest.bodyCallBack`](https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadTaskRequest.kt#L18) uses arbitrary boundary in this situation. It leads to
```
PUT http://localhost/ HTTP/1.1
Accept-Encoding: compress;q=0.5, gzip;q=1.0
Content-Type: application/x-www-form-urlencoded
Cache-Control: no-cache
Pragma: no-cache
User-Agent: Java/1.8.0_172
Host: localhost:8080
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: keep-alive
Content-Length: 209263

--164e202956e
Content-Disposition: form-data; name="lat"
Content-Type: text/plain
```
Content-Type is urlencoded, but body is multipart/form-data, and library user should set boundary manually. Is it intended?

### LINKED ISSUE #479: Improve multipart features for httpPost and httpUpload
# Feature Request

With the current API, some scenarios of multipart interaction are not friendly and require what imho are hacks.

## Description
I am working with CMS API and there's a API call with requires uploading a document alongside metadata in JSON.
This is easily done in curl with:
```cUrl
curl "https://api.hostname/v1/documents" \
-X POST \
-H 'Content-Type: multipart/form-data' \
-H "Authorization: Bearer $token" \
-H "Repository-Id: $repo_id" \
-F metadata="{
    \"description\": \"Test document\",
    \"categories\": [{...}]
	};type=application/json" \
-F file=@file_path
```

I only got this to work with the following code

```kotlin

    val (request, response, result) = url.httpUpload(
//            parameters = listOf("metadata" to """{"description":"test document from Kotlin"};type=application/json""")
    )
            .dataParts { request, url ->
                listOf(
                        DataPart(File("src/main/resources/rust-logo-blk.svg"), "file"),
                        DataPart(File("src/main/resources/doc.json"), "metadata","application/json")
                )
            }
            .header("Authorization" to "Bearer $token")
            .header("Repository-Id" to repoId)
            .responseString()
```

As you can see this requires storing the JSON in a file. That's because I could not find any way to set the Content-Type for the part other than using `DataPart`. Using `parameters` as seen in the commented line sets it as `text/plain` causing the service to return an error.

I also tried something like

```kotlin
    val (request, response, result) = url.httpPost(listOf(
            "metadata" to """{"description":"test document from Kotlin"}""",
            "file" to DataPart(File("resources/rust-logo-blk.svg"))))
            .header("Content-Type" to "multipart/form-data")
            .header("Authorization" to "Bearer $token")
            .header("Repo-Id" to repoId)
            .responseString()
```

## Proposed Solution
Imho, the naming `upload` is confusing for this cases but is very useful for simple 1 file uploads. I'd advocate to:
1. Create a httpMultipart similar to httpPost that offers the ability to pass any kind of object, not only files and set the Content-Type. Or allow htttPost to set Content-Type to multipart so that it manages the format internally (a la curl).
2. Extend DataPart to support any data (String, Inputstream, ...)

Probably, with only point 2 would be enough to support a cleaner method with httpUpload.

## Alternatives I've considered
Current solution does not allow to generate metadata Json on-the fly, which is not valid for a dynamic system.
Also, the `dataParts` closure element seems cumbersome to me and does not help readability. I prefer simpler/cleaner `httpPost` syntax. But this is a more personal thing.

## Additional context
I noticed that when no Content-Type is set in the DataPart, the header is added empty, maybe this could be just removed.

### LINKED ISSUE #513: [fuel-core] Overwrite header in case of Content-Type: multipart/form-data
# Feature Request

## Description
This feature solves a problem I've had when constructing a Multipart/form-data request.

If one sets the Content-Type header to multipart/form-data manually (done in the snippet below) this request wont work properly, since the boundry value in the header, pointing out the file, is missing.

If the manually set content-type is removed, the request works fine.

```
Fuel.upload(path = baseUrl, method = Method.POST)
                .header(mutableMapOf(
                    "apikey" to key,
                    "Accept" to "application/json",
                    "Cache-Control" to "no-cache"
                    "Content-Type" to "multipart/form-data"
                ))
                .dataParts { _, _ -> listOf(DataPart(file, "file", "image/jpeg")) }
                .responseJson { _, _, result ->
                    result.fold(
                            success = { data -> cb(data.content) },
                            failure = { error ->
                                println("An error of type ${error.exception} happened: ${error.message}")
                                cb("")
                            }
                    )
                }
```

## Proposed Solution
Fuel should in my opinion in the case of such a specific request, *overwrite* the header even though a user beforehand set it manually, since the request otherwise wont work at all.

### PR DESCRIPTION
**Pull request by SleeplessByte**
## Description

Based on RFC7578: https://tools.ietf.org/html/rfc7578

This is a breaking change, but mostly deprecates current functions as error.

Removes all the extranous methods and bookkeeping from UploadRequest and UploadBody, and delegates those to the DataPart. This also allows for InlineDataPart that yield a file on the server and removes automatic naming for dataparts (_1, _2), because the correct way to put multiple files under the same name is to give them... the same name!

Fixes #344
Fixes #390
Fixes #399
Fixes #479
Fixes #513

## Type of change

Check all that apply

- [X] Bug fix (a non-breaking change which fixes an issue)
- [X] New feature (a non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactoring (a change which changes the current internal or external interface)
- [X] This change requires a documentation update

## How Has This Been Tested?

Tests have been updated and new ones have been added

> Note that due to deprecation this _DOES_ generate compiler warnings, but these are expected.

## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation, if necessary
- [ ] My changes generate no new compiler warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Inspect the bytecode viewer, including reasoning why

### CODE REVIEW

File: fuel/README.md [hunk 1]
```diff
@@ -390,60 +390,69 @@ Fuel supports multipart uploads using the `.upload()` feature. You can turn _any
...
 
 ```kotlin
-val formData = listOf("Email" to "mail@example.com", "Name" to "Joe Smith" )
-
-Fuel.upload("/post", param = formData)
-    .dataParts { request, url -> listOf<DataPart>() }
+Fuel.upload("/post")
+    .add(
+        FileDataPart(File("myfile.json"), name = "file", fileName="contents.json"),
+        InlineDataPart(myinlinecontent, name = "metadata", fileName="metadata.json", contentType = "application/json")
```
Comment (markGilchrist) [inline fuel/README.md]: [nits] should `myinlinecontent` be `myInlineContent` ?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt [hunk 2]
```diff
@@ -1,9 +1,223 @@
...
+ *    // Content-Disposition: "form-data; name=field-name[]; filename=baz.json"
+ *    // Content-Type: "application/json"
+ *    //
+ *    // <file-contents baz.json>
+ *
+ *  @param file [File] the source file
+ *  @param name [String] the field name, defaults to the file name without extension
+ *  @param fileName [String] the remote file name, or null
+ *  @param contentType [String] the content type of the file contents; defaults to a guess using [guessContentType]
+ *  @param contentDisposition [String] defaults to form-data with name and filename set, unless filename is null
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt]: @SleeplessByte  `the Shakespeare of kDocs` !

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadBody.kt [hunk 3]
```diff
@@ -1,19 +1,28 @@
...
-import java.net.URLConnection
 import java.nio.charset.Charset
 
+internal class BoundaryMissing(request: UploadRequest) : FuelError(
+    IllegalArgumentException(
+        "The Request is missing the boundary parameter in its Content-Type.\n\n" +
+        "This can happen if you manually overwrite the Content-Type but forget to set a boundary. The boundary is \n" +
+        "normally set automatically when you call \"request.upload()\". Remove manually setting the Content-Type or \n" +
+        "add the boundary parameter to the Content-Type for this request: \n\n" +
+            "\trequest.header(Headers.ContentType, \"multipart/form-data; boundary=custom-boundary\")"
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadBody.kt]: great

File: fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt [hunk 4]
```diff
@@ -290,12 +291,8 @@ class BlockingRequestTest : MockHttpTestCase() {
 
         val (request, response, data) =
                 manager.upload(mock.path("upload"), parameters = listOf("foo" to "bar", "foo1" to "bar1"))
-                        .source { _, _ ->
-                            val dir = System.getProperty("user.dir")
-                            val currentDir = File(dir, "src/test/assets")
-                            File(currentDir, "lorem_ipsum_long.tmp")
-                        }
-                        .responseString()
+                    .add { FileDataPart(File(File(System.getProperty("user.dir"), "src/test/assets"), "lorem_ipsum_long.tmp")) }
```
Comment (kittinunf) [inline fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt]: We have `File(File` here sir!

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt [hunk 5]
```diff
@@ -1,9 +1,223 @@
...
+ */
+data class FileDataPart(
     val file: File,
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt]: [Question] As this is a `FileDataPart`, do you think we can add another API in such a way that we can skip passing `File` object altogether for the sake of ease of use? For example, In the test the usage could be

```kotlin
request.add(FileDataPart(System.getProperty("user.dir").resolve("src/test/assets"), "lorem_ipsum_long.tmp"))
```

I think the usage of passing a "String" or a path that indicates the location of the file should be the most ubiquitous?

File: fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt (hunk 4 above)
Comment (SleeplessByte) [reply fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt]: Hah. That's a woooopsiedaisy.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt (hunk 5 above)
Comment (kittinunf) [reply fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt]: Ah, I think that is a reasonable approach! 👍

File: fuel/src/test/kotlin/com/github/kittinunf/fuel/RequestPathStringConvertibleExtensionTest.kt [hunk 6]
```diff
@@ -151,11 +152,7 @@ class RequestPathStringConvertibleExtensionTest : MockHttpTestCase() {
 
         val (request, response, result) = PathStringConvertibleImpl(mock.path("http-upload"))
             .httpUpload()
-            .source { _, _ ->
-                val dir = System.getProperty("user.dir")
-                val currentDir = File(dir, "src/test/assets")
-                File(currentDir, "lorem_ipsum_long.tmp")
-            }
+            .add { FileDataPart(File(File(System.getProperty("user.dir"), "src/test/assets"), "lorem_ipsum_long.tmp")) }
```
Comment (kittinunf) [inline fuel/src/test/kotlin/com/github/kittinunf/fuel/RequestPathStringConvertibleExtensionTest.kt]: Here too sir!

File: fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt (hunk 4 above)
Comment (SleeplessByte) [reply fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt]: I changed it to

```kotlin
FileDataPart.from(
  File(System.getProperty("user.dir")).resolve("/src/test/assets"), 
  "lorem_ipsum_long.tmp"
)
```

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt (hunk 5 above)
Comment (SleeplessByte) [reply fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt]: This now works.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt [hunk 7]
```diff
@@ -1,9 +1,259 @@
...
+         * @return [FileDataPart] the DataPart
+         */
+        fun from(directory: File, filename: String, name: String? = null, remoteFileName: String? = filename, contentType: String? = null): DataPart {
+            val file = File(directory, filename)
+            return FileDataPart(
+                file = file,
+                name = name ?: file.nameWithoutExtension,
+                filename = remoteFileName,
+                contentType = contentType ?: guessContentType(file)
+            )
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/DataPart.kt]: **[ktlint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Unexpected blank line(s) before "}"

File: fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt [hunk 8]
```diff
@@ -288,14 +289,11 @@ class BlockingRequestTest : MockHttpTestCase() {
...
+        val path = File(System.getProperty("user.dir"), "/src/test/assets").resolve("lorem_ipsum_long.tmp").path
         val (request, response, data) =
                 manager.upload(mock.path("upload"), parameters = listOf("foo" to "bar", "foo1" to "bar1"))
-                        .source { _, _ ->
-                            val dir = System.getProperty("user.dir")
-                            val currentDir = File(dir, "src/test/assets")
-                            File(currentDir, "lorem_ipsum_long.tmp")
-                        }
-                        .responseString()
+                    .add { FileDataPart.from(path,  "lorem_ipsum_long.tmp") }
```
Comment (kittinunf) [inline fuel/src/test/kotlin/com/github/kittinunf/fuel/BlockingRequestTest.kt]: **[ktlint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Unnecessary space(s)

**Pull request merged**